include/SDL3/SDL_filesystem.h

```diff
@@ -236,6 +236,125 @@ typedef enum
+typedef int (SDLCALL *SDL_EnumerateCallback)(void *userdata, SDL_FSops *fs, const char *origdir, const char *fname);
+
+/**
+ * This is the directory operation interface, much like SDL_RWops is the filestream
+ * operation interface.  SDL provides a platform-independent way to
+ * access these that operate on the native filesystem, but the app and other libraries
+ * are welcome to provide their own implementations of any data that needs a simple
+ * tree-shaped hierarchy that looks a little like a filesystem.
+ *
+ * This struct is public so apps and libraries can provide their own implementations,
+ * but one should not call these function pointers directly to use an implementation;
+ * please call the SDL_FS* functions instead, as there is common code there that needs
+ * to run in addition to the implementation's function pointers.
+ */
+struct SDL_FSops
+{
+    /**
+     * Open a file in the filesystem.
+     *
+     * You do not necessarily have to use this, if you know you're working from
+     * a real filesystem and just want to enumerate paths and such.
+     *
+     * The caller should not close the filesystem while the returned RWops is still open.
+     *
+     * \param fs The SDL_FSops object to use.
+     * \param path the path in the SDL_FSops to open
+     * \param mode a fopen-style "mode" string of how to access the path.  !!! FIXME: maybe just offer read/write/append enums?
+     * \return a RWops for the path specified.
+     */
+    SDL_RWops *(SDLCALL * open)(SDL_FSops *fs, const char *path, const char *mode);
+
+    /* list all files in a directory. Each file is passed through a callback until we're done or the callback returns <= 0. */
+    int (SDLCALL * enumerate)(SDL_FSops *fs, const char *path, SDL_EnumerateCallback cb, void *userdata);
+
+    /* delete/unlink/rmdir a path. Will not remove non-empty directories! */
+    int (SDLCALL * remove)(SDL_FSops *fs, const char *path);
+
+    /* Create a directory. */
+    int (SDLCALL * mkdir)(SDL_FSops *fs, const char *path);
+
+    /* Obtain basic path metadata. */
+    int (SDLCALL * stat)(SDL_FSops *fs, const char *path, SDL_Stat *stat);
+
+    /* Close the filesystem and free its resources. */
+    void (SDLCALL * closefs)(SDL_FSops *fs);
+
+    /* Used by the SDL_FSops for private data. Don't touch! */
+    void *opaque;
+
+    /* Everything has properties! :) */
+    SDL_PropertiesID props;
+};
+
+
+/*
+ * Create an SDL_FSops that represents the native filesystem, treating `basedir` as
+ * the root of the tree. A NULL basedir is valid and gives you the actual root of the
+ * native filesystem.
+ *
+ * Enumerating root on Windows will give you available drive letters like they were directories,
+ * almost everything else assumes the root is a Unix "/" thing.
+ *
+ * Destroy this when done with SDL_DestroyFilesystem().
+ */
+extern DECLSPEC SDL_FSops * SDLCALL SDL_CreateFilesystem(const char *basedir);
```

Comment:
Can you "../.." out of the basedir?

Reply:
Currently, yes.

In PhysicsFS, we check all path strings to make sure there are no ".." directories included, so if someone used the library for mod support, the mod can't break out its confined space, but that wasn't something I was trying to do here, and it might be useful to allow relative paths that move outside the specified basedir anyhow.

Reply:
I think one of the first things people will do with a basedir is use it for mod support. We should prevent `..` out of the root.